Add Docker-based build script

skorokithakis opened the pull request:
Pertaining to issue #16722, I'm making this PR to add a simple Docker-based build script to the repo. It enables anyone to build the firmware by just doing:

```
git clone https://github.com/ArduPilot/ardupilot.git
./docker_build.sh omnibusf4
```

I have needed to add `gcc-arm-none-eabi` to the Dockerfile, otherwise some boards wouldn't build.

Feel free to close if this is unwanted.

amilcarlucas commented:
Please follow the commit message policy defined at: https://ardupilot.org/dev/docs/submitting-patches-back-to-master.html#preparing-commits

skorokithakis commented:
Done, thanks @amilcarlucas. I can't change the branch without opening a new PR, hopefully that won't be a problem.

khancyr commented on Dockerfile:
```diff
@@ -35,6 +35,9 @@ RUN echo "if [ -d \"\$HOME/.local/bin\" ] ; then\nPATH=\"\$HOME/.local/bin:\$PAT
 # Set the buildlogs directory into /tmp as other directory aren't accessible
 ENV BUILDLOGS=/tmp/buildlogs
 
+RUN sudo apt-get update && sudo apt-get install -y gcc-arm-none-eabi
```
this is wrong ... we already have the right toolchain in the container and don't want to use Ubuntu one's

skorokithakis replied:
It definitely wouldn't build for me without this, maybe there were path issues? It wouldn't find the compiler.

skorokithakis replied:
You can try removing that line and running the script to reproduce.

khancyr replied:
Just enter the container and you will be able to compile fine

skorokithakis replied:
Hmm, so it is a path issue? I wanted to be able to run one command and get the output, otherwise it's too many steps. I can look into adding the proper path additions to compile, though.

skorokithakis replied:
Well you are correct, running `bash` and then `./waf build` inside the container works fine... I've reverted this line and will investigate further.

khancyr commented on docker_build.sh:
```diff
@@ -0,0 +1,17 @@
+git submodule update --init --recursive
+
+docker build . -t ardupilot
+docker run --rm -it -v "$(pwd)":/ardupilot ardupilot:latest ./waf configure --board="$BOARD"
```
something is wrong with this invocation as the dockerfile provided is definitively able to compile all boards

khancyr replied:
and the docker_build.sh  should be in Tools/scripts

skorokithakis replied:
Hmm, very odd... I added this because people usually want to compile a single board, but that can be done when the board argument is omitted.

khancyr replied:
Humm, I wonder if the mistake isn't just a path thing. Normally, the stm toolchain add to the .bashrc, so the mistake may be that doing that way the .bashrc isn't loaded

skorokithakis replied:
That's very likely, I'll investigate.

skorokithakis commented:
You were correct, @khancyr, in the end it was just required to set the PATH. I have added the relevant environment variable in the Dockerfile and updated the script. It now builds fine with a single command.

I have also added `intelhex` to the Dockerfile to enable generation of `.hex` firmware, which is important to me.

peterbarker commented on Tools/scripts/docker_build.sh:
```diff
@@ -0,0 +1,20 @@
+
+set -euo pipefail
+
+if [ $# -ne 1 ]
+  then
```
```suggestion
if [ $# -ne 1 ]; then
```

for no particular reason apart from that's what we do most places.

peterbarker commented on Tools/scripts/docker_build.sh:
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
One of the few places I've ever seen `/usr/bin/env bash` :-)

skorokithakis replied:
When you work with Nix for a bit, you pick this up :P

peterbarker reviewed (changes requested):
I really like this concept.

skorokithakis commented on Tools/scripts/docker_build.sh:
```diff
@@ -0,0 +1,20 @@
+cd "$(git rev-parse --show-toplevel)"
+git submodule update --init --recursive
+
+PATH='/home/ardupilot/.local/bin:/usr/lib/ccache:/ardupilot/Tools/autotest:/opt/gcc-arm-none-eabi-6-2017-q2-update/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
```
Hmm, this is an outdated line, it does only tack stuff onto PATH now.

How do you mean implicit? I added it here because otherwise `waf` stuff won't work in `docker run` unless you run `bash` and do things manually there.

skorokithakis replied:
Oops, I forgot this when debugging, removed.

skorokithakis commented on Tools/scripts/docker_build.sh:
```diff
@@ -0,0 +1,20 @@
+
+if [ $# -ne 1 ]
+  then
+    echo "No board supplied, run as ./docker_build.sh <board name> or ./docker_build.sh list"
```
This is a bit of a hack, `list` isn't a valid board name so it prints a list of boards. `./docker_build.sh foobar` would do the same, really.

khancyr commented on Dockerfile:
```diff
@@ -35,8 +35,11 @@ RUN echo "if [ -d \"\$HOME/.local/bin\" ] ; then\nPATH=\"\$HOME/.local/bin:\$PAT
     && sudo rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
 
 ENV CCACHE_MAXSIZE=1G
+ENV PATH="/home/ardupilot/.local/bin:/usr/lib/ccache:/ardupilot/Tools/autotest:/opt/gcc-arm-none-eabi-6-2017-q2-update/bin:${PATH}"
```
I don't want to maintain that. so the best way would be to modify the installation script to create .ardupilot.bash file with all our PATH modification. Then on the dockerside load the .ardupilot.bash file in the .bashrc for non login shell and in the .bash_profile for the login shell (like docker run cmd)

khancyr commented on Tools/scripts/docker_build.sh:
```diff
@@ -0,0 +1,17 @@
+BOARD=$1
+
+cd "$(git rev-parse --show-toplevel)"
+git submodule update --init --recursive
```
we don't want this by default

skorokithakis replied:
This PR is meant to enable one-step building of firmware for people who don't know about development. This way, all they need is Docker and they can build the firmware. Taking steps out of the script defeats is purpose, but feel free to close this PR if you don't want a one-step builder.

khancyr replied:
updating the submodule is part of our instructions to setup the repository correctly, so that isn't an issue.

The main problem by hidding the submodule update there is that you hide it and then people won't understand why it fails. Moreover it will also prevent to work on some submodule as it will always erase the changes.

skorokithakis replied:
> updating the submodule is part of our instructions to setup the repository correctly, so that isn't an issue.

Certainly, but then it's not "one-step". Having a "few steps" way is fine, and my main problem was that the PATH wasn't getting set correctly so the Docker image would only build if you entered the bash shell inside the container, but I also wanted a one-step way to build.

> The main problem by hidding the submodule update there is that you hide it and then people won't understand why it fails. Moreover it will also prevent to work on some submodule as it will always erase the changes.

I agree, but the users who know what "submodule" (or even "git") means wouldn't need to use this script, or would better be able to debug it if it breaks. This is specifically so the instructions can say "Run these commands: 'git clone; ./docker_build.sh' and the firmware will be in the `build/` directory".

It's fine if you think that, for example, novices shouldn't be trying to build, but currently the instructions are basically 3-4 steps. To me, reducing that number to 1 is worth it, 2 kind of isn't, so I'd be perfectly okay if we didn't merge this in the main repo and I could post a note somewhere with it, for new users.

I can also remove the submodule update, if you think that's worthwhile, but then it's just a script that contains two Docker commands, and might as well be `docker run bash -c "waf configure; waf build"`.

peterbarker replied:
I'm really a bit confused as to what happened here.  My understanding is that this was a stand-alone script which would let people get started developing ardupilot more easily than they currently do.  I'm a bit disappointed this was simply closed without a defining reason as to why it was a bad idea.

Note that in regards to the submodules - we should have patches in master which mean that they are magically initialised on first build - so it really ought not to have been an issue here.

skorokithakis replied:
I closed it because, on one hand, there didn't seem to be much want for such a one-step script, and, on the other, it "works for me" and I didn't want to spend however many hours trying to figure out how to make bash load the appropriate PATH (though I'm mystified as to why it doesn't, it runs fine if you start bash in the container).
